src/libutil/filesystem.cpp

```diff
@@ -514,6 +523,182 @@ Filesystem::open (std::ofstream &stream, string_view path,
+
+static int
+ios_open_mode_to_oflag(std::ios_base::openmode mode)
+{
+    int f = 0;
+    if (mode & std::ios_base::in) {
+        f |= _O_RDONLY;
+    }
+    if (mode & std::ios_base::out) {
+        f |= _O_WRONLY;
+        f |= _O_CREAT;
+        if (mode & std::ios_base::app) {
+            f |= _O_APPEND;
+        }
+        if (mode & std::ios_base::trunc) {
+            f |= _O_TRUNC;
+        }
+    }
+    if (mode & std::ios_base::binary) {
+        f |= _O_BINARY;
+    } else {
+        f |= _O_TEXT;
+    }
+    return f;
+}
+
+static std::istream*
+open_ifstream_impl(string_view path,
+                   std::ios_base::openmode mode)
+{
+    std::wstring wpath = Strutil::utf8_to_utf16(path);
+    int fd;
+    int oflag = ios_open_mode_to_oflag(mode);
+    errno_t errcode = _wsopen_s(&fd, wpath.c_str(), oflag, _SH_DENYNO, _S_IREAD | _S_IWRITE);
+    if (errcode != 0) {
+        return 0;
+    }
+    __gnu_cxx::stdio_filebuf<char>* buffer = new __gnu_cxx::stdio_filebuf<char>(fd, mode, 1);
+    if (!buffer) {
+        return 0;
+    }
+    return new std::istream(buffer);
+}
+
+static std::ostream*
+open_ofstream_impl(string_view path,
+                   std::ios_base::openmode mode)
+{
+    
+    std::wstring wpath = Strutil::utf8_to_utf16(path);
+    int fd;
+    int oflag = ios_open_mode_to_oflag(mode);
+    errno_t errcode = _wsopen_s(&fd, wpath.c_str(), oflag, _SH_DENYNO, _S_IREAD | _S_IWRITE);
+    if (errcode != 0) {
+        return 0;
+    }
+    __gnu_cxx::stdio_filebuf<char>* buffer = new __gnu_cxx::stdio_filebuf<char>(fd, mode, 1);
+    if (!buffer) {
+        return 0;
+    }
+    return new std::ostream(buffer);
+}
+#else // MSVC or Unix
+
+#ifdef _WIN32
+#ifndef _MSC_VER_
+#error "open_ifstream_impl only supports GCC or MSVC"
+#endif
+#endif
+
+static std::ifstream*
+open_ifstream_impl(string_view path,
+                   std::ios_base::openmode mode)
+{
+    
+#ifdef _WIN32
+    std::wstring wpath = Strutil::utf8_to_utf16(path);
+#endif
+    std::ifstream *ret = new std::ifstream();
+    if (!ret) {
+        return 0;
+    }
+    try {
+        ret->open(
+#ifdef _WIN32
+                  wpath.c_str(),
+#else
+                  path.c_str(),
+#endif
+                  mode);
+    } catch (const std::exception & e) {
+        delete ret;
+        return 0;
+    }
+    
+    if (!*ret) {
+        delete ret;
+        return 0;
+    }
+    
+    return ret;
+} // open_ifstream_impl
+
+static std::ofstream*
+open_ofstream_impl(string_view path,
+                   std::ios_base::openmode mode)
+{
+#ifdef _WIN32
+    std::wstring wpath = Strutil::utf8_to_utf16(path);
+#endif
+    std::ofstream *ret = new std::ofstream();
+    if (!ret) {
+        return 0;
+    }
+    try {
+        ret->open(
+#ifdef _WIN32
+                  wpath.c_str(),
+#else
+                  path.c_str(),
+#endif
+                  mode);
+    } catch (const std::exception & e) {
```

Comment:
Just IMHO, I believe that this would be clearer as

```
#ifdef _WIN32
    ret->open (wpath.c_str(), mode);
#else 
    ret->open (path.c_str(), mode);
#endif 
```

rather than having the #ifdef split up within the argument list of the function call, particularly when the whole function call can fit on one line.